### PR TITLE
Implement header chunk read

### LINKED ISSUE #4: Implement chunk or stream read to reduce the number of `read` call
Right now when reading the header we have to loop through each FID to get each FID metadata. 
https://github.com/mapbox/COGDumper/blob/dfc6b9b56879c7116f522518ed37617d570acba1/cogdumper/cog_tiles.py#L221-L222

While this permit to read only part we need to determine what will be the offsets for the data and mask part, this is also not efficient, resulting in 10s of small `read` calls. 

```
cogdumper s3 --bucket mapbox --key playground/vincent/y.tif --xyz 0 0 0
Read Header
read 0 3
read 4 7
read 8 9
read 10 251
read 516 771
read 260 515
read 826 967
read 250 253

Read IFD
read 1378 1379
read 1380 1549
read 1808 2063
read 1552 1807
read 1548 1551

Read IFD
read 2064 2065
read 2066 2259
read 2332 2395
read 2268 2331
read 2450 2591
read 2258 2261

Read IFD
read 2592 2593
read 2594 2787
read 2812 2827
read 2796 2811
read 2882 3023
read 2786 2789

Read IFD
read 3024 3025
read 3026 3219
read 3282 3423
read 3218 3221

Read IFD
read 3424 3425
read 3426 3619
read 3682 3823
read 3618 3621

Read IFD
read 3824 3825
read 3826 4019
read 4082 4223
read 4018 4021

Read IFD
read 4224 4225
read 4226 4419
read 4482 4623
read 4418 4421

Read IFD
read 4624 4625
read 4626 4795
read 4862 4925
read 4798 4861
read 4794 4797

Read IFD
read 4926 4927
read 4928 5097
read 5116 5131
read 5100 5115
read 5096 5099

Read IFD
read 5132 5133
read 5134 5303
read 5302 5305

Read IFD
read 5306 5307
read 5308 5477
read 5476 5479

Read IFD
read 5480 5481
read 5482 5651
read 5650 5653

Read IFD
read 5654 5655
read 5656 5825
read 5824 5827

Read IFD
read 883331 908790
read 2341520 2341571
```

I don't remember exactly but it seems that GDAL is reading the first 16ko of the file and then determine all it needs. I think applying the same logic could be nice.

cc @normanb @sgillies

### PR DESCRIPTION
**Pull request by vincentsarago**
closes #4 
This PR try to implement `chunk` reading for header retrieval to limit the number of `read` calls

### To Do
Right now I'm reading the first 8000 bytes of a file and decode the header. It works for normal TIFF but not for BIGTIFF. @normanb do you now if there is a size limit for GeoTIFF headers ?

### CODE REVIEW

File: cogdumper/cog_tiles.py [hunk 1]
```diff
@@ -50,7 +50,10 @@ def __init__(self, reader):
         self._image_ifds = []
         self._mask_ifds = []
 
-    def ifds(self):
+        # Read header here using stream
+        self._read_header()
+
+    def _ifds(self, buff):
```
Comment (vincentsarago) [inline cogdumper/cog_tiles.py]: Small change, made `ifds` a private method.

**Comment by normanb** [comment]
afaik there isn't a limit on the header size for either tiff or bigtiff, as the header can have a flexible number of  tags and these tags can have a different number of values.  Picking a large enough chunk size and being prepared to go back for more seems to be an approach. Ideally chunk size should be configurable.

**Comment by vincentsarago** [comment]
I update most of the change and it should work now. I need to add more test to make sure everything is fine 

cc @normanb

**Comment by vincentsarago** [comment]
@normanb I'm not sure I can make test coverage better. I think this PR is ready for final review. 
Just noting that by default we fetch the first `1024` bytes 
https://github.com/mapbox/COGDumper/blob/35de35bd4bfa929a72674755e17521065eb6c950/cogdumper/cog_tiles.py#L217
while GDAL fetch the first 16k. We could change that to make it more GDAL like and/or are a environment variable to define the size of the first fetch.

**Comment by vincentsarago** [comment]
@normanb in https://github.com/mapbox/COGDumper/pull/5/commits/7d9849258fa86649b52cea3cc3f106ea266ae72c I added a environment variable `COG_INGESTED_BYTES_AT_OPEN` used to configure the size of the first read as GDAL 2.3 is doing http://gdal.org/gdal_virtual_file_systems.html#gdal_virtual_file_systems_vsicurl. 

I also added a logging option to 👀  the number and size read. 

```
cogdumper file --file cog_deflate.tif --xyz 0 0 0  -v
INFO:cogdumper.filedumper:Reading bytes: 0 to 16383
INFO:cogdumper.filedumper:Reading bytes: 11434145 to 11831775
```

I guess if this is 👌 you can do a final review and merge it :-)

**Comment by normanb** [comment]
+1 Thanks @vincentsarago for these changes, it greatly improves performance.

**Comment by vincentsarago** [comment]
@normanb should we ship a new release with the change implemented here ?

**Pull request merged**